### PR TITLE
Typescript support for ALL templates

### PR DESCRIPTION
**Pull request by tobiasqueck**
This PR contains the first iteration of TypeScript support in our writer modules.

* `ui5-application-writer`
   * new advanced option `typescript: boolean` added that if set to true generates a valid UI5 TypeScript application
* `fiori-freestyle-writer`
   * new typescript option from ui5-app also exposed here
   * TS support added for all templates
   * fixed flpPreview in templates
* `fe-fpm-writer`
   * typescript flag added and implemented for all functionality
* `fiori-elements-writer`
   * new typescript option from ui5-app also exposed here
   * new template fpm added that utilizes the new typescript support in the `fe-fpm-writer`
   * enabled `sapux` flag as default

### CODE REVIEW

File: packages/fiori-elements-writer/src/data/defaults.ts
Comment (devinea) [inline packages/fiori-elements-writer/src/data/defaults.ts:95]: Need to consider if this would impact CAP where FE does not have sapux = true.

Comment (tobiasqueck) [reply packages/fiori-elements-writer/src/data/defaults.ts:95]: I would consider it an exceptional case, so the Fiori generator would have to set `sapux: false` explicitly for CAP.

File: packages/fiori-elements-writer/templates/common/add/webapp/Component.ts
Comment (IainSAP) [inline packages/fiori-elements-writer/templates/common/add/webapp/Component.ts:1]: Isnt `Component.js/ts` provided by ui5-application-writer?

File: packages/fiori-elements-writer/src/data/defaults.ts
Comment (tobiasqueck) [reply packages/fiori-elements-writer/src/data/defaults.ts:95]: The flag indicates that the generated project is supported by the Fiori tools. We want people to be able to use the tools with things we know we can support, so the flag should always have been true. Nevertheless, if the caller explicitly sets it to `false` it is respected.
So for the CAP case, where we need it differently, I expect that when this is integrated by the Fiori generator, that someone now sets it explicitly to false. Current version of the fe write is `0.4.7`, this one changes it to `0.5.0` so it is aligned with semver (minor change below 1)

File: packages/fiori-elements-writer/src/index.ts
Comment (IainSAP) [inline packages/fiori-elements-writer/src/index.ts:81]: Im not sure this is necessary as we already have validation for service version here: https://github.com/SAP/open-ux-tools/pull/626/files#diff-a51a19d6fa0b94939e84e2a4eedd1b602cae40c2077c22ecce59ee537f5ccbcfR40 . Can we use the same approach for the new template also?

File: packages/ui5-application-writer/src/index.ts
Comment (IainSAP) [inline packages/ui5-application-writer/src/index.ts:90]: We already write these files here: https://github.com/SAP/open-ux-tools/pull/626/files#diff-1bff8e3813a4755fbee9efa4f5c4d3cbb9cf7c0b8bbf32e83399a394320e1060R47 
Can we consolidate the writing of these files to one place instead of overwriting?

File: packages/fiori-elements-writer/src/index.ts
Comment (tobiasqueck) [reply packages/fiori-elements-writer/src/index.ts]: Thanks, I have removed the check and added a test case validating it with https://github.com/SAP/open-ux-tools/pull/626/commits/56127e442f97065ab01e0d4568863ef3d8f18e6b

**Review by IainSAP — DISMISSED** [review]
Thx @tobiasqueck

-  Comments addressed
-  Tests added to cover

**Review by IainSAP — DISMISSED** [review]
Re-approving following merge from main.

**Review by IainSAP — DISMISSED** [review]

**Review by devinea — COMMENTED** [review]
Was it intentional to not provide a TS version of `packages/ui5-application-writer/templates/optional/loadReuseLibs/webapp/utils/locate-reuse-libs.js` ?

File: packages/fiori-freestyle-writer/test/worklist.test.ts
Comment (devinea) [inline packages/fiori-freestyle-writer/test/worklist.test.ts:143]: This test is not using typescript. 
```
appOptions:{
                    typescript: true
                }
```
Adding above. Navigation to object page fails as there bug in packages/fiori-freestyle-writer/test/test-output/worklistTemplate/worklist_ts/webapp/controller/Object.controller.ts 

````

        var oResourceBundle = this.getResourceBundle(),
            oObject = oView.getBindingContext().getObject(),
            sObjectId = oObject.,
            sObjectName = oObject.;
````
VS Code shows various problems for the file
Screenshot 2022-08-25 at 17 07 35

**Review by devinea — CHANGES_REQUESTED** [review]

**Comment by devinea** [comment]
@tobiasqueck Generally it would good to have a TS test project test-output for each JS project test-output. There are gaps currently especially in FE.

**Comment by tobiasqueck** [comment]
> Was it intentional to not provide a TS version of packages/ui5-application-writer/templates/optional/loadReuseLibs/webapp/utils/locate-reuse-libs.js ?

Yes. I assumed that this is a file that nobody is supposed to modify and it is also nothing that is part of the actual app comonent.

File: packages/fiori-freestyle-writer/test/worklist.test.ts
Comment (tobiasqueck) [reply packages/fiori-freestyle-writer/test/worklist.test.ts:143]: That is odd. I played with the output of all tests when using the `UX_DEBUG` variable. I will have a look at it again.
Thanks for spotting it!

Comment (tobiasqueck) [reply packages/fiori-freestyle-writer/test/worklist.test.ts]: Test fixed with https://github.com/SAP/open-ux-tools/pull/626/commits/a0bf107ec87c049fcdee46eba2bd140967a918ea and object page issue fixed with https://github.com/SAP/open-ux-tools/pull/626/commits/9c5d5d3ad74d54c8d95c528391ff91271dc21de0

Comment (tobiasqueck) [reply packages/fiori-freestyle-writer/test/worklist.test.ts]: > VS Code shows various problems for the file

Improved with https://github.com/SAP/open-ux-tools/pull/626/commits/ad1a11535ce97c041714ad8ec7e7cb5f5d790909.

**Comment by tobiasqueck** [comment]
> Generally it would good to have a TS test project test-output for each JS project test-output. There are gaps currently especially in FE.

I'd disagree. TS is one additional option that can be used, like enabling reuse libs. We also don't have a test for every combination with reuse libs. Especially in FE, adding it now for each template would not test anything new that isn't test in the FPM TS template. In addition, the `fe-fpm-writer` has an integration test that covers a LROP app with a custom FPM page and lots of extensions. So, I really believe we have covered our bases here.

**Comment by devinea** [comment]
> > Was it intentional to not provide a TS version of packages/ui5-application-writer/templates/optional/loadReuseLibs/webapp/utils/locate-reuse-libs.js ?
> 
> Yes. I assumed that this is a file that nobody is supposed to modify and it is also nothing that is part of the actual app component.

Yeah OK. Fair enough.

File: packages/ui5-application-writer/templates/optional/typescript/package.json
Comment (devinea) [inline packages/ui5-application-writer/templates/optional/typescript/package.json:3]: This script on its own is not that useful. it should be integrated into build and or preview scripts. 

When I run `npm run ts-typecheck` in `/packages/fiori-freestyle-writer/test/test-output/listDetailTemplate/listdetail-ts`.

```
Found 124 errors in 8 files.

Errors  Files
     2  webapp/Component.ts:45
     8  webapp/controller/App.controller.ts:10
     8  webapp/controller/BaseController.ts:5
    48  webapp/controller/Detail.controller.ts:4
     8  webapp/controller/ErrorHandler.ts:22
    35  webapp/controller/List.controller.ts:11
    12  webapp/controller/ListSelector.ts:21
     3  webapp/controller/NotFound.controller.ts:8
````
All these errors should be fixed (and in other floorplans) to be a true TS project.

File: packages/fiori-freestyle-writer/test/worklist.test.ts
Comment (devinea) [reply packages/fiori-freestyle-writer/test/worklist.test.ts]: Thanks

File: packages/ui5-application-writer/templates/optional/typescript/package.json
Comment (tobiasqueck) [reply packages/ui5-application-writer/templates/optional/typescript/package.json:3]: @devinea  Would it be ok for you to accept these errors for now, and we merge this new minor version. I will start right away with a new branch on creating a patch fixing these errors and some others issues I saw in the js templates.
I thought about fixing these here right away but the PR is big enough already.

**Review by devinea — APPROVED** [review]
Thanks  @tobiasqueck for the updates. As discussed there are a few type check errors 
in `packages/fiori-freestyle-writer/test/test-output/listDetailTemplate/listdetail-ts`
`npm i && npm run ts-typecheck`
Screenshot 2022-08-26 at 16 53 07

These are to be addressed in a follow-up PR.

Also noticed that `packages/fiori-freestyle-writer/test/test-output/basicTemplate/basic_typescript` `npm run start-mock` fails with `server:ux-fe-mockserver TypeError: Cannot read property '_attributes' of undefined` which seems to be more related to the mockserver and the metadata used in the test.

**Comment by tobiasqueck** [comment]
@devinea thanks for the approval. I will merge the PR as soon as I get a 2nd approval. I will not close the issue and will look at the open ts errors and the issue you found with the mockdata asap.

File: packages/ui5-application-writer/templates/optional/typescript/package.json
Comment (tobiasqueck) [reply packages/ui5-application-writer/templates/optional/typescript/package.json:3]: Resolving the conversation as discussed.

**Review by IainSAP — APPROVED** [review]
Approving, as follow-up PR to be merged soon :)

**Pull request merged**
